### ISSUE TITLE
Convert theano weights to tensorflow format

### ISSUE THREAD

**Issue by lhk** [fchollet/keras#5374]
I would like to use this pretrained model: https://gist.github.com/baraldilorenzo/07d7802847aaad0a35d3

It is in theano layout, my code depends on tensorflow image dimension ordering.

There is a guide on converting weights between the formats: https://github.com/fchollet/keras/wiki/Converting-convolution-kernels-from-Theano-to-TensorFlow-and-vice-versa

But this seems broken. In the section to convert theano to tensorflow, the first instruction is to load the weights into the tensorflow model.

> Keras backend should be TensorFlow in this case.
>First, load the Theano-trained weights into your TensorFlow model:
`model.load_weights('my_weights_theano.h5')`

This raises an exception, the weight layouts would be incompatible.
And if the load_weights function would take theano weights for a tensorflow model, there wouldn't be any need to convert them.

I took a look at the convert_kernel function. To see, wether I could do the necessary steps myself: https://github.com/fchollet/keras/blob/master/keras/utils/np_utils.py
The code is rather simple - I don't understand why the guide makes use of a tensorflow session. That seems unnecessary.

I've copied the code from the pretrained model to create a model with tensorflow layer. This just meant changing the input shape and the backend.image_dim_ordering before adding any Convolutions.
Then I used this loop:

`model` is the original model, created from the code I linked at the beginning.
`model_tensorflow` is the exact same model, but with tensorflow layout.
```
for i in range(len(model.layers)):
    layer_theano=model.layers[i]
    layer_tensorflow=model_tensorflow.layers[i]
    
    if layer_theano.__class__.__name__ in ['Convolution1D', 'Convolution2D', 'Convolution3D', 'AtrousConvolution2D']:
        weights_theano=layer_theano.get_weights()
        
        kernel=weights_theano[0]
        bias=weights_theano[1]
        
        converted_kernel=convert_kernel(kernel, "th")
        converted_kernel=converted_kernel.transpose((3,2,1,0))
        
        weights_tensorflow=[converted_kernel, bias]

        layer_tensorflow.set_weights(weights_tensorflow)
        
    else:
        layer_tensorflow.set_weights(layer_theano.get_weights())
```

In the original code, there is a testcase: Prediction ran on the image of a cat.
I've downloaded a cat image and tried the testcase with the original model: 285.
The converted model predicts 585.

I don't know whether 285 is the correct label for a cat, but even if it isn't, the two models should be broken in the same way, I would expect the same prediction.

What is the correct way of converting weights between models ?

**Comment by titu1994** [comment]
@lhk I was having the same trouble for quite some time. Finally, I was able to transfer the weights in quite a complicated way, but it works well. 

I've made a script which auto translates theano backend (th dim ordering) trained model weights directly into the other 3 possible combinations of backend / dim ordering.

[The script is available here](https://github.com/titu1994/Keras-Classification-Models/blob/62ab38512c6ef0fe4a0f30a00d33643f453cf12e/weight_conversion_theano.py)

Since it is such a complicated process, it has a few instructions on how to use it, and you need to manually edit the script in some places (like importing your own script to build a model, creating a theano and tensorflow model and specifying the path to the weight files for theano backend th dim ordering model).

Also, it runs on the tensorflow backend only for now. But for all that, it returns 3 folders with the correctly translated weights.

Note: The weight shuffle algorithm is from Kent Sommers (@kentsommer). I haven't tested it yet, but I think it works correctly. 

In a few days (I have exams coming up, so not much time), I hope to translate the script to work with models which were trained on Tensorflow with tf dim ordering, and get it to auto translate the weights to Theano.

**Comment by lhk** [comment]
That looks fantastic. 
I can't try it right now, but it's the first thing I'll do in the morning. 

Thank you very much.

**Comment by titu1994** [comment]
@lhk No problem. This was a huge problem for me for several months, so it's nice to finally have a working answer.

**Comment by titu1994** [comment]
@fchollet Would it be possible to edit the wiki to point to the above mentioned script? It simplifies the procedure quite a bit.

**Comment by kentsommer** [comment]
Nice job @titu1994! 

Again @lhk, the row_shuffle algorithm has not been tested so it is very possible that it is incorrect, however, it matches to the best of my knowledge what was done for the VGG16 and VGG19 models in the Keras applications examples.

**Comment by lhk** [comment]
Unfortunately, it does seem to be incorrect. 

I took the code for vgg16 in theano and changed the input shape to fit tensorflow. Then the code can be executed to create a model with tensorflow backend. 

I provided the theano and tensorflow models, as well as the weights to your script. 
Then I reran the vgg16 script, the if `__name__=="__main__"` part. The predictions from the original and the converted models are different.

**Comment by titu1994** [comment]
Hmm it seems that the only time it converts weights in the wrong manner is when there are two or more dense layers after the final convolutional layers. 

Does the model you are converting have more than one dense layer after the final convolutional layer ? If not, the weights are correctly converting. 

I tested it out by checking each and every layer weight with already translated weight provided by fchollet for the ResNet model with np.allclose.

I will try and modify my checking script to see if I can make it print out the row translation scheme. 

Sorry for the trouble.

**Comment by lhk** [comment]
Yes, it does have more than one dense layer.

**Comment by titu1994** [comment]
Ok. I've modified my checking script to now print out the row shuffles. 

It will take quite some time (it checks 25k rows by 25k rows) but once it finished I'll be able to see where the wrong translations are occuring.

**Comment by lhk** [comment]
Thank you very much

**Comment by titu1994** [comment]
Sorry for the delay, I had classes. 

Surprisingly, an 8-core i7 processor with 16 GB RAM with joblib parallelization took nearly 1 hour to process 25k rows by 25 cols. In any case, I have the results now and a simple diff shows me the changes I have to make.

Will push the changes and notify when done.

**Comment by titu1994** [comment]
@lhk well Found the bug.

@kentsommer The algorithm you used is exactly matching the row shuffles done by fchollet, but the way you were updating the shuffled weights was a little faulty.

The corrected version of your code : 
```
def shuffle_rows(original_w, nb_last_conv, nb_rows_dense):
    converted_w = np.zeros(original_w.shape)
    count = 0
    for index in range(original_w.shape[0]):
        if (index % nb_last_conv) == 0 and index != 0:
            count += 1
        new_index = ((index % nb_last_conv) * nb_rows_dense) + count
        print("index from " + str(index) + " -> " + str(new_index))
        converted_w[index] = original_w[new_index] # Here converted[i] needs to update to original[new_index]. Not the other way around. 

    return converted_w
```

I've pushed the fixes to both of my scripts now. They finally work completely for all types of models, and match the weights provided by fchollet.

**Comment by kentsommer** [comment]
@titu1994 
Good catch :+1: 

Might want to confirm that this swap holds from TensorFlow -> Theano (I looked at this direction for my original code) as well as Theano -> TensorFlow (I'm assuming you tested this direction). Awesome work making this all into one script! Definitely useful :1st_place_medal:

**Comment by titu1994** [comment]
There's another script for conversion from Tensorflow to Theano, since the folder names change and the dim ordering changes and also I'd have to interpret the weight shape myself to determine which transfer was occuring. 

So I simply leave it to the user to determine the correct script. There's weight_transfer_theano.py for conversion from Theano kernels TH dim to all other and weight_transfer_tensorflow.py for converting from Tensorflow kernels TF dim ordering to all others.

**Comment by smarinka** [comment]
@titu1994 ,  your script worked perfectly for me

**Comment by titu1994** [comment]
@smarinka I've not tested this for weights generated by Keras 2. 

Keras 2 may not even need weight weight shuffling anymore as the changelog suggests that weights will work regardless of which backend and dim ordering you use.

**Comment by fenollp** [comment]
I have been trying to convert a h5+json model trained on Theano backend (on Keras 1), to Keras 2 on Tensorflow 1 backend and indeed just for this purpose weights work regardless.

However I want my Tensorflow weights in the format pb+txt [as found here](https://github.com/tensorflow/tensorflow/blob/208192074d7d19f6b51724bb06fd4c2a143649e7/tensorflow/contrib/ios_examples/camera/CameraExampleViewController.mm#L25-L36).
Any idea how to do that? [h5+json here](https://github.com/JostineHo/real-time_emotion_analyzer)

Note to @titu1994: your script crashed on this model with
```
...snip...
Done tf-kernels-th-dim model.h5
Nb layers :  18
Saved layer 1 : convolution2d_10
Saved layer 2 : convolution2d_11
Saved layer 3 : convolution2d_12
Saved layer 4 : maxpooling2d_4
Saved layer 5 : convolution2d_13
Saved layer 6 : convolution2d_14
Saved layer 7 : convolution2d_15
Saved layer 8 : maxpooling2d_5
Saved layer 9 : convolution2d_16
Saved layer 10 : convolution2d_17
Saved layer 11 : convolution2d_18
Saved layer 12 : maxpooling2d_6
Saved layer 13 : flatten_2
Traceback (most recent call last):
  File "weight_conversion_theano.py", line 128, in <module>
    nb_rows_dense_layer = weights[0].shape[0] // nb_last_conv
ZeroDivisionError: integer division or modulo by zero
```

Code used to load model (not sure I got this right):
```python3
from keras.models import model_from_json
with open('model.json') as fd:
    jsn = model_from_json(fd.read()).to_json()  # converts from Keras 1

K.set_image_dim_ordering('th')
th_dim_model = model_from_json(jsn)
th_dim_model.load_weights('model.h5')

K.set_image_dim_ordering('tf')
tf_dim_model = model_from_json(jsn)

model_weights = ['model.h5']
```
Ideas?

**Comment by PaulScemama** [comment]
@titu1994 thanks for the great script. I had been working on a deep learning algorithm for CRISPR guide-RNA design that was originally coded with Theano backend. I needed to convert to R, but the Theano backend was not working (may not be supported anymore in R/keras?). 

FYI, my input was a 2D one-hot encoding matrix that represented a DNA sequence, and thus used Convolutions in the model. Therefore, the original script did not correspond (as I believe the script is meant for higher-dimensional image input). I modified the script for it to work with 2D one-hot encoding matrices if you'd be interested in implementing that feature.
